### DISCUSSION TITLE
Coefficient field (material properties) reconstruction through MOOSE - Wave Equation

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#18547] [Q&A General] [answered]
Hi

I am working on two types of the inverse problem (reconstruction of the material properties) that involves solving the wave equation in the **frequency domain** for multiple right-hand sides. This problem is well known in geophysics (seismic inversion) and in biomedical imaging (Ultrasound imaging). Typically, this problem involve thousands of forward solves (FE simulation) as a results of sweeping over all frequency content and all sources (right hand sides) at each iteration to the optimization (reconstruction). 

Although I am just starting learning MOOSE and trying to switch to this new mindset, I can imagine how powerful MOOSE is especially for multiphysics simulation. However, I am still have some issues in deciding whether MOOSE is a suitable choice for my problems. 

To be specific, the optimum framework that I am looking for is preferably have the following:

1- Can handle complex arithmetics (needed for solving the wave equation in the frequency domain)
2- Ability to modify the integration rule for specific region/regions of the domain (e.g. mid point integration), this is needed to model a variant of the perfectly matched layer (PMDL) that acts as wave absorber. 
3- Abitility to update material properties after each optimization iteration in runtime without the need for updating the input.i file after each iteration. 

So, if you could please give some pointers on the previous 3 points and some pointer on how to model an arbitrary, completely heterogenous 3D domain (i.e each finite element in the mesh has different material property). 

I really appreciate your response.

**Comment by GiudGiud** [comment, marked as answer]
Hello

We can definitely do 3. this is very routine. People can define material properties that are function of time, space, temperature, you name it. Their values can be different in every element, and can be recomputed as often as you want. 

So you would want to use a different element type or quadrature rule in parts of the domain for 2? This is also possible, though more involved if this is not part of the ones we have already implemented in the framework. If you have an equation for that we may be able to give more details. 

I ll let people more expert about the wave equation reply about 1.

Guillaume

**Reply by hugary1995** [reply]
I think op wants different time integrators for different subdomains, which is currently not supported.

**Reply by GiudGiud** [reply]
Ah ok
You could do a multiapp approach for this. It would be like a separate calculation that you iterate to convergence.

**Reply by aaelmeli** [reply]
Thank you, Guillaume and Gary, for the response.

With respect to the integration part (point 2), let's, for example, consider an elasticity problem in 2D, and consider the bilinear Q4 finite element, the full integration requires 4 Gauss Points, this is already there and I have no problem with it. 
Now, for some cases, for instance, when the problem suffers from volumetric locking, we need to do "selective reduced integration" where the volumetric strain is evaluated by only integrating at  **one** Gauss Point in the middle of the Q4 element, whereas the shear strain is evaluated by full integration. 

This is a very helpful technique when modeling elastic half-spaces and wave absorbers. In addition, it can be used for modeling soft tissues where soft tissues are known to be incompressible.

**Reply by aaelmeli** [reply]
> I think op wants different time integrators for different subdomains, which is currently not supported.

I do not think I would need different time integrators. The problem that I solve is in the frequency domain. So, we would have many steady-state problems (corresponds to each frequency in the spectrum) as opposed to marching in time.

**Reply by hugary1995** [reply]
Ah sorry I misunderstood. Yeah, we do volumetric locking correction routinely. Mostly B-bar and F-bar.

**Reply by hugary1995** [reply]
Regarding your point one, I know libMesh definitely is able to solve complex eigenvalue problems, so there's no reason moose can't do that. I haven't done that myself, I think @fdkong or maybe @roystgnr know more about that.

**Reply by roystgnr** [reply]
The trouble on the libMesh side is that, because PETSc (and thereby SLEPc) always forced users to choose between real and complex PetscScalar at configure time, libMesh never bothered to add runtime-selectable support for real and complex either.  For users who were writing individual complex eigenproblem solver applications this wasn't a hassle, but for frameworks like Moose?  Building a whole copy of Moose against an --enable-complex libMesh is what you'd have to do, and I'm not sure whether people actually can do that or not.
